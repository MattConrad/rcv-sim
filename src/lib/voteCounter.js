const getBestRemainingChoice = (vote, losers) => {
  let position = 1;

  while(vote[position]) {
    if (!losers.includes(vote[position])) return vote[position];
    position = position + 1;
  }
};

const getTotal = (segments) => (
  Object.values(segments).reduce((total, voteCount) => total + voteCount)
);

const getLoser = (results) => {
  let loser;
  let min;
  Object.keys(results).map((key) => {
    const total = getTotal(results[key])
    if (!min || total < min) {
      loser = key;
      min = total;
    }
  })
  return loser;
}

const getVoteSegments = (votes, losers) => {
  return votes.reduce((counts, vote) => {
    const topChoice = vote[1]; 
    const bestRemainingChoice = getBestRemainingChoice(vote, losers);
    if (bestRemainingChoice) {
      if (!counts[bestRemainingChoice]) counts[bestRemainingChoice] = {};
      if (!counts[bestRemainingChoice][topChoice]) counts[bestRemainingChoice][topChoice] = 0;
      counts[bestRemainingChoice][topChoice] = counts[bestRemainingChoice][topChoice] + 1;
    }
    return counts;
  }, { });
};

const getVoteSegmentsForRound = (votes, round) => {
  const losers = [];
  let thisRound = round;
  let segments = {};
  while(thisRound) {
    segments = getVoteSegments(votes, losers);
    losers.push(getLoser(segments));
    thisRound = thisRound - 1;
  }
  return segments;
}

export const getResultsForRound = (votes, round) => {
  const segments = getVoteSegmentsForRound(votes, round);
  const results = { candidateTotals: {}, segments };
  let overallTotal = 0;
  Object.keys(segments).forEach((key) => {
    const total = getTotal(segments[key]);
    results.candidateTotals[key] = total;
    overallTotal = overallTotal + total;
  });

  results.overallTotal = overallTotal;

  return results;
}

