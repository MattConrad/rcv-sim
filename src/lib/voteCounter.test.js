import { getResults } from './voteCounter';

const earlyWin = [
  { 1: 'D', 2: 'B', 3: 'C', 4: 'A', 5: 'E' },
  { 1: 'D', 2: 'C', 3: 'B', 4: 'A', 5: 'E' },
  { 1: 'D', 2: 'E', 3: 'C', 4: 'A', 5: 'B' },
  { 1: 'C', 2: 'B', 3: 'A', 4: 'D', 5: 'E' },
  { 1: 'E', 2: 'B', 3: 'C', 4: 'D', 5: 'A' },
  { 1: 'A', 2: 'D', 3: 'C', 4: 'B', 5: 'E' },
  { 1: 'D', 2: 'A', 3: 'C', 4: 'E', 5: 'B' }
];

const balanced = [
  { 1: 'B', 2: 'D', 3: 'C', 4: 'A', 5: 'E' },
  { 1: 'D', 2: 'C', 3: 'B', 4: 'A', 5: 'E' },
  { 1: 'E', 2: 'A', 3: 'C', 4: 'D', 5: 'B' },
  { 1: 'C', 2: 'B', 3: 'A', 4: 'D', 5: 'E' },
  { 1: 'E', 2: 'B', 3: 'C', 4: 'D', 5: 'A' },
  { 1: 'A', 2: 'D', 3: 'C', 4: 'B', 5: 'E' },
  { 1: 'D', 2: 'A', 3: 'C', 4: 'E', 5: 'B' }
];

const candidates = ['A', 'B', 'C', 'D', 'E'];

it('should run only round 0 and return results', () => {
  const results = getResults(earlyWin, candidates);

  expect(results.length).toEqual(1);
  expect(results[0].winner).toEqual('D');
  expect(results).toMatchSnapshot('early win');
});

it('should run multiple rounds with the winner on the last only', () => {
  const results = getResults(balanced, candidates);
  const firstRound = results[0];
  const secondRound = results[1];
  const lastRound = results[results.length - 1];

  expect(results.length).toEqual(3);
  expect(firstRound.winner).toBeUndefined();
  expect(secondRound.winner).toBeUndefined();
  expect(lastRound.winner).toEqual('D');
  expect(results).toMatchSnapshot('balanced win');
});

it('should provide full segment lists for each candidate', () => {
  const results = getResults(balanced, candidates);
  const firstRound = results[0];
  const secondRound = results[1];
  const lastRound = results[results.length - 1];

  expect(results.length).toEqual(3);
  expect(firstRound.winner).toBeUndefined();
  expect(secondRound.winner).toBeUndefined();
  expect(lastRound.winner).toEqual('D');
  expect(results).toMatchSnapshot('balanced win');
});
