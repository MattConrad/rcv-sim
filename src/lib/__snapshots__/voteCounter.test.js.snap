// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balanced win 1`] = `
Array [
  Object {
    "A": Map {
      "A" => 1,
    },
    "B": Map {
      "B" => 1,
    },
    "C": Map {
      "C" => 1,
    },
    "D": Map {
      "D" => 2,
    },
    "E": Map {
      "E" => 2,
    },
    "loser": "A",
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "B": Map {
      "B" => 1,
      "A" => 0,
    },
    "C": Map {
      "C" => 1,
      "A" => 0,
    },
    "D": Map {
      "D" => 2,
      "A" => 1,
    },
    "E": Map {
      "E" => 2,
      "A" => 0,
    },
    "loser": "B",
    "validVoteCount": 7,
    "winner": undefined,
  },
  Object {
    "C": Map {
      "C" => 1,
      "A" => 0,
      "B" => 0,
    },
    "D": Map {
      "D" => 2,
      "A" => 1,
      "B" => 1,
    },
    "E": Map {
      "E" => 2,
      "A" => 0,
      "B" => 0,
    },
    "loser": "C",
    "validVoteCount": 7,
    "winner": "D",
  },
]
`;

exports[`early win 1`] = `
Array [
  Object {
    "A": Map {
      "A" => 1,
    },
    "B": Map {
      "B" => 0,
    },
    "C": Map {
      "C" => 1,
    },
    "D": Map {
      "D" => 4,
    },
    "E": Map {
      "E" => 1,
    },
    "loser": "C",
    "validVoteCount": 7,
    "winner": "D",
  },
]
`;
